go/libraries/utils/async/ring_buffer.go

```diff
@@ -81,13 +85,16 @@ func (rb *RingBuffer) Close() error {
 }
 
 // Push will add a new item to the RingBuffer and signal a go routine waiting inside Pop that new data is available
-func (rb *RingBuffer) Push(item interface{}) error {
+func (rb *RingBuffer) Push(item interface{}, epoch int) error {
 	rb.cond.L.Lock()
 	defer rb.cond.L.Unlock()
 
 	if rb.closed {
 		return os.ErrClosed
 	}
+	if epoch != rb.epoch {
+		return nil
```

Comment:
If this is general purpose, may want to return `ErrBadEpoch`.  In the case of index lookups can do a check and ignore `ErrBadEpoch`

Reply:
Good idea. Added.